File: lib/services/ocr_space_service.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:http/http.dart' as http;
import 'package:main_project_files/services/api_service.dart';


class OCRSpaceService {
  static const String apiKey = "K83335691088957"; // replace

  static Future<String> extractTextFromImage(Uint8List imageBytes) async {
    final uri = Uri.parse("https://api.ocr.space/parse/image");

    final request = http.MultipartRequest("POST", uri)
      ..fields["language"] = "eng"
      ..fields["OCREngine"] = "2"
      ..fields["scale"] = "true"
      ..fields["isTable"] = "false"
      ..headers["apikey"] = apiKey
      ..files.add(
        http.MultipartFile.fromBytes(
          "file",
          imageBytes,
          filename: "scan.jpg",
        ),
      );

    final response = await request.send();
    final respString = await response.stream.bytesToString();

    try {
      final data = jsonDecode(respString);

      if (data["IsErroredOnProcessing"] == true) {
        print("OCR.Space error: ${data["ErrorMessage"]}");
        return "";
      }

      final text = data["ParsedResults"]?[0]?["ParsedText"] ?? "";
      return text.trim();
    } catch (e) {
      print("OCR.Space parse error: $e");
      return "";
    }
  }
}

File: src/services/ocrSpace.service.js
// // src/services/ocrSpace.service.js
// import axios from "axios";
// import FormData from "form-data";

// export async function extractTextFromImage(imageBuffer) {
//   const apiKey = process.env.OCR_SPACE_API_KEY;
//   if (!apiKey) {
//     throw new Error("OCR_SPACE_API_KEY missing in env");
//   }

//   const form = new FormData();
//   form.append("file", imageBuffer, "scan.jpg");
//   form.append("language", "eng");
//   form.append("OCREngine", "2");
//   form.append("scale", "true");
//   form.append("isTable", "false");

//   const response = await axios.post(
//     "https://api.ocr.space/parse/image",
//     form,
//     {
//       headers: {
//         ...form.getHeaders(),
//         apikey: apiKey
//       },
//       maxBodyLength: Infinity
//     }
//   );

//   const data = response.data;

//   if (data.IsErroredOnProcessing) {
//     console.error("OCR.Space error:", data.ErrorMessage);
//     return "";
//   }

//   return data.ParsedResults?.[0]?.ParsedText?.trim() || "";
// }

/* =========================================================
   ocrSpace.service.js — v2.0
   Base64 upload (more reliable than buffer)
   Retry on timeout, better OCR params
   ========================================================= */

import axios from "axios";

const OCR_API_URL = "https://api.ocr.space/parse/image";
const TIMEOUT_MS  = 30000; // 30s
const MAX_RETRIES = 2;

export async function extractTextFromImage(imageBuffer) {
  const apiKey = process.env.OCR_SPACE_API_KEY;
  if (!apiKey) throw new Error("OCR_SPACE_API_KEY missing in env");

  // Convert buffer to base64 data URI
  const base64 = imageBuffer.toString("base64");
  const dataURI = `data:image/jpeg;base64,${base64}`;

  const payload = new URLSearchParams();
  payload.append("base64Image",        dataURI);
  payload.append("language",           "eng");
  payload.append("OCREngine",          "2");
  payload.append("scale",              "true");
  payload.append("isTable",            "false");
  payload.append("isOverlayRequired",  "false");
  payload.append("detectOrientation",  "true");  // handles rotated labels
  payload.append("filetype",           "jpg");

  let lastError;

  for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
    try {
      const response = await axios.post(OCR_API_URL, payload, {
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
          "apikey":        apiKey,
        },
        timeout: TIMEOUT_MS,
      });

      const data = response.data;

      if (data.IsErroredOnProcessing) {
        const msg = Array.isArray(data.ErrorMessage)
          ? data.ErrorMessage.join(", ")
          : data.ErrorMessage;
        throw new Error(`OCR.Space error: ${msg}`);
      }

      const text = data.ParsedResults?.[0]?.ParsedText?.trim() ?? "";
      return text;

    } catch (err) {
      lastError = err;
      const isTimeout = err.code === "ECONNABORTED" || err.message?.includes("timeout");
      if (!isTimeout || attempt === MAX_RETRIES) throw err;
      // Wait briefly before retry
      await new Promise(r => setTimeout(r, 1500));
    }
  }

  throw lastError;
}